File: valoreal/api_models.py
"""Typed client-facing schemas for the ValScores API."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class APIModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ErrorDetail(APIModel):
    code: str
    message: str


class ErrorResponse(APIModel):
    error: ErrorDetail


class TimelineMatchResponse(APIModel):
    id: int = Field(ge=1)
    team1: str
    team2: str
    team1_score: int | None
    team2_score: int | None
    team1_round_score: int | None
    team2_round_score: int | None
    status: Literal["upcoming", "live", "finished"]
    start_time: datetime | None
    is_live: bool
    is_finished: bool


class MapStatResponse(APIModel):
    game_id: str
    map_number: int = Field(ge=0)
    map_name: str
    team1_round_score: int | None
    team2_round_score: int | None


class PlayerStatResponse(APIModel):
    name: str
    team: str
    role: str
    acs: int
    kd: float
    adr: int
    kills: int
    deaths: int
    assists: int
    plus_minus: str
    kast: str
    first_kills: int
    first_deaths: int


class MatchStatsInfoResponse(APIModel):
    team1: str
    team2: str
    map_vetoes: list[str]
    selected_game_id: str
    maps: list[MapStatResponse]


class MatchStatsResponse(APIModel):
    match_info: MatchStatsInfoResponse
    team1_roster: list[PlayerStatResponse]
    team2_roster: list[PlayerStatResponse]


class RefreshFailureResponse(APIModel):
    stage: str
    error: str
    source: str | None = None
    vlr_match_id: str | None = None
    url: str | None = None
    card_index: int | None = Field(default=None, ge=1)
    href: str | None = None


class RefreshResponse(APIModel):
    status: Literal["success", "partial", "failed"]
    matches_examined: int = Field(ge=0)
    inserted: int = Field(ge=0)
    updated: int = Field(ge=0)
    unchanged: int = Field(ge=0)
    details_refreshed: int = Field(ge=0)
    failure_count: int = Field(ge=0)
    failures: list[RefreshFailureResponse]

File: valoreal/APIModels.swift
import Foundation

struct MatchStatsResponse: Decodable {
    let match_info: MatchStatsInfo
    let team1_roster: [PlayerStatInfo]
    let team2_roster: [PlayerStatInfo]
}

struct MatchStatsInfo: Decodable {
    let team1: String
    let team2: String
    let map_vetoes: [String]
    let selected_game_id: String
    let maps: [MapStatInfo]
}

struct MapStatInfo: Decodable, Identifiable {
    var id: String { game_id }
    let game_id: String
    let map_number: Int
    let map_name: String
    let team1_round_score: Int?
    let team2_round_score: Int?
}

struct PlayerStatInfo: Decodable, Identifiable {
    var id: String { "\(team)-\(name)" }
    let name: String
    let team: String
    let role: String
    let acs: Int
    let kd: Double
    let adr: Int
    let kills: Int
    let deaths: Int
    let assists: Int
    let plus_minus: String
    let kast: String
    let first_kills: Int
    let first_deaths: Int
}

struct RefreshResponse: Decodable {
    let status: String
    let matches_examined: Int
    let inserted: Int
    let updated: Int
    let unchanged: Int
    let details_refreshed: Int
    let failure_count: Int
    let failures: [RefreshFailure]
}

struct RefreshFailure: Decodable {
    let stage: String
    let error: String
    let source: String?
    let vlr_match_id: String?
    let url: String?
    let card_index: Int?
    let href: String?
}
